copy: improve behavior of non-atomic COPY

yuzefovich opened the pull request:
This commit audits the behavior of non-atomic COPY. As a reminder, in such a setup we create a separate txn for each batch of rows, and the responsibility of committing these txns lies with the COPY state machine. To further complicate the story, we have a retry loop around each batch of rows, where we give each batch up to 5 times to be processed in case it is hitting retryable errors.

One problem that this commit addresses is how we reset the buffered data. Previously, we would do the reset right after executing the INSERT pipeline. In some cases, that pipeline results in the txn being auto committed, so we're good then; however, if the INSERTs cannot auto commit, then it is possible that we will fail to commit the txn in the COPY state machine, but by that point we will have reset the buffered data, effectively silently losing it. On the second retry attempt for this batch we'd have zero rows to insert, so in absence of other problems, we could end up proceeding with the COPY without inserting some of the data. This commit addresses this scenario by resetting the buffered state only after we know that the INSERTs have been committed.

However, I believe previously it wouldn't be possible to hit this silent data loss scenario because of a second issue this commit addresses. In particular, when the COPY state machine fails to commit its txn, it attempts to roll it back, and previously, regardless of whether the rollback succeeded or not, we would always short-circuit execution of the `cleanup` function returned by `preparePlannerForCopy` so that we'd end up keeping the same txn, that is in finalized state and couldn't be used anymore. Thus, on the next retry attempt we'd always get a non-retrible error failing the COPY overall. This commit addresses this issue by not short-circuiting the execution of that deferred method which creates a fresh txn for the next retry attempt.

It also now makes it so that if a txn rollback fails, then that non-retriable error is returned to the client, failing the COPY. This seems like a sane thing to me because at that point we're kind of in an undefined state. However, I could also see a benefit of proceeding still since we're in non-atomic COPY scenario and perhaps the user wants COPY to succeed even if it results in somewhat corrupted inserted data.

Kudos to Nathan for identifying these two problems.

As an additional minor change it fixes the number of inserted rows that could previously be inflated in case non-atomic COPY had internal retries (although, again, due to the second issue described above, this was hidden).

Epic: None

Release note (bug fix): CockroachDB can now retry more retryable errors transparently when performing non-atomic COPY command.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/116512)
<!-- Reviewable:end -->

yuzefovich committed:
copy: short-circuit earlier

This commit makes the logic in `insertRowsInternal` short-circuit the
execution earlier in case we have no rows to insert. In particular, this
allows us to avoid possibly creating a new txn that won't be used at
all.

Release note: None

yuzefovich committed:
copy: improve behavior of non-atomic COPY

This commit audits the behavior of non-atomic COPY. As a reminder, in
such a setup we create a separate txn for each batch of rows, and
the responsibility of committing these txns lies with the COPY state
machine. To further complicate the story, we have a retry loop around
each batch of rows, where we give each batch up to 5 times to be
processed in case it is hitting retryable errors.

One problem that this commit addresses is how we reset the buffered
data. Previously, we would do the reset right after executing the INSERT
pipeline. In some cases, that pipeline results in the txn being auto
committed, so we're good then; however, if the INSERTs cannot auto
commit, then it is possible that we will fail to commit the txn in the
COPY state machine, but by that point we will have reset the buffered
data, effectively silently losing it. On the second retry attempt for
this batch we'd have zero rows to insert, so in absence of other
problems, we could end up proceeding with the COPY without inserting
some of the data. This commit addresses this scenario by resetting the
buffered state only after we know that the INSERTs have been committed.

However, I believe previously it wouldn't be possible to hit this silent
data loss scenario because of a second issue this commit addresses. In
particular, when the COPY state machine fails to commit its txn, it
attempts to roll it back, and previously, regardless of whether the
rollback succeeded or not, we would always short-circuit execution of
the `cleanup` function returned by `preparePlannerForCopy` so that we'd
end up keeping the same txn, that is in finalized state and couldn't be
used anymore. Thus, on the next retry attempt we'd always get
a non-retrible error failing the COPY overall. This commit addresses
this issue by not short-circuiting the execution of that deferred
method which creates a fresh txn for the next retry attempt.

It also now makes it so that if a txn rollback fails, then that
non-retriable error is returned to the client, failing the COPY. This
seems like a sane thing to me because at that point we're kind of in an
undefined state. However, I could also see a benefit of proceeding still
since we're in non-atomic COPY scenario and perhaps the user wants COPY
to succeed even if it results in somewhat corrupted inserted data.

Kudos to Nathan for identifying these two problems.

As an additional minor change it fixes the number of inserted rows that
could previously be inflated in case non-atomic COPY had internal
retries (although, again, due to the second issue described above, this
was hidden).

Release note (bug fix): CockroachDB can now retry more retryable errors
transparently when performing non-atomic COPY command.

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116512)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2, @nvanbenschoten, and @rafiss)

___
*[`-- commits` line 45 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-NlfMHnl0qz33Z44Ke-Q):*
I'm interested in feedback on this paragraph. Before this change when we failed to roll back the txn, we would proceed (unless we failed to commit the txn so we'd end up with a finalized txn that would fail the COPY on the next retry). Currently, each fail to roll back fails the COPY.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
copy: add a session variable for the number of retries per batch

Previously, for every batch of the non-atomic COPY would get retried up
to 5 times (meaning there would be at most 6 attempts). This commit
introduces a session variable that can control this number - this might
come handy in the future as an additional knob to make COPY succeed
(e.g. in the context of DMS).

Release note: None

yuzefovich commented:
Friendly ping @rafiss (I think you know COPY state machine transitions best) and @nvanbenschoten (since you looked closely into this - if you have time).

rafiss reviewed (changes requested):
i like this change a lot! thanks. my comments are relatively minor

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-:-NmCH-O0288c2C7yFseF:b-lt101p)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2, @nvanbenschoten, and @yuzefovich)

___
*[`-- commits` line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-NmCEwJF5vM9lh311Vnq):*
i agree with the change described in the last paragraph

but i'm not sure if this PR should be backported. was there an external need to backport it? it seems like it could bring in behavioral changes, and also, it's a somewhat sizable refactor so has some risk in its own right.
___
*[`pkg/sql/copy_from.go` line 1034 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-NmCFcmd3i1GYegy8pR4:-NmCFcmd3i1GYegy8pR5:buw44s0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bf21961ac3a11c8d7a98d9aa3058138a8688847/pkg/sql/copy_from.go#L1034)):*
> ```Go
> 							// the data, so we return this non-retriable error.
> 							log.Warningf(ctx, "non-atomic COPY couldn't roll back its txn: %v", rollbackErr)
> 							return rollbackErr
> ```

nit: maybe we could use `return errors.Wrap(rollbackErr, "non-atomic COPY couldn't roll back its txn")`. (and same below)
___
*[`pkg/sql/copy_from.go` line 1038 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-NmCGC1K07iWWI652bjn:-NmCGC1K07iWWI652bjo:b-8ko74q) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bf21961ac3a11c8d7a98d9aa3058138a8688847/pkg/sql/copy_from.go#L1038)):*
> ```Go
> 						// The rollback succeeded, so we can simply attempt to
> 						// retry this batch, after having prepared a fresh txn
> 						// below.
> ```

would this be the place to adjust the `insertedRows` counter?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116512)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2, @nvanbenschoten, and @rafiss)

___
*[`-- commits` line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-NmCEwJF5vM9lh311Vnq):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i agree with the change described in the last paragraph

but i'm not sure if this PR should be backported. was there an external need to backport it? it seems like it could bring in behavioral changes, and also, it's a somewhat sizable refactor so has some risk in its own right.
</blockquote></details>

This change was prompted by https://github.com/cockroachlabs/support/issues/2765. I agree that this change is not without its risks. The main reason for why I'd like to get this backported (at least to 23.2.x) is that it should reduce the likelihood that non-atomic COPY returns retryable errors to the client. In the last couple of months Queries team had like 3-4 L2 tickets that sought more guidance on how to make COPY succeed. Addressing the second issue mentioned in the commit message should make those types of escalations less likely.
___
*[`pkg/sql/copy_from.go` line 1038 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-NmCGC1K07iWWI652bjn:-NmCvSQqDAM4DFwFg2-W:b-wplidg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bf21961ac3a11c8d7a98d9aa3058138a8688847/pkg/sql/copy_from.go#L1038)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

would this be the place to adjust the `insertedRows` counter?
</blockquote></details>

I think we want to updated `insertedRows` right when we're reseting the rows. Included the fix into the second commit.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Friendly ping @rafiss @michae2

michae2 reviewed (changes requested):
Sorry for the long wait. I just have one question about the first commit. Nice job fixing this!

I'm not sure if we should backport to 23.1. Definitely in favor of backporting to 23.2.

> Reviewed 10 of 10 files at r9, 5 of 5 files at r10, 7 of 7 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-:-NnozMyl3XPy12AzrjLm:b-kifufi)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @rafiss, and @yuzefovich)

___
*[`pkg/sql/copy_from.go` line 1090 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-NnohyZeDe9oGKd7iYVi:-NnohyZeDe9oGKd7iYVj:b-8104z6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbd78f458ab20759553a6220ec589023c61e32eb/pkg/sql/copy_from.go#L1090)):*
> ```Go
> 		return nil
> 	}
> 	cleanup := c.p.preparePlannerForCopy(ctx, &c.txnOpt, finalBatch, c.implicitTxn)
> ```

Don't we still need to call `preparePlannerForCopy` if `finalBatch` is true, just in case `CopyFromAtomicEnabled` and `implicitTxn` are both true and this final empty batch should be the one to commit the transaction?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Ack, let's backport only to 23.2.1.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-:-Nnpu54w0EngCPHWzanW:b-d4gdea)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2, @nvanbenschoten, and @rafiss)

___
*[`pkg/sql/copy_from.go` line 1090 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-NnohyZeDe9oGKd7iYVi:-Nnpvpn903yGXeE1CbvD:b-f9xi6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbd78f458ab20759553a6220ec589023c61e32eb/pkg/sql/copy_from.go#L1090)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Don't we still need to call `preparePlannerForCopy` if `finalBatch` is true, just in case `CopyFromAtomicEnabled` and `implicitTxn` are both true and this final empty batch should be the one to commit the transaction?
</blockquote></details>

Hm, let me think.

If we have non-atomic copy, then we create a separate txn for each batch, and each txn is either auto-committed during the execution of the INSERT flow or in the deferred `cleanup` function returned by `preparePlannerForCopy` - in other words, it is effectively guaranteed (except for error cases) that if we called `preparePlannerForCopy` and it created a txn, then that txn will be committed.

If we have atomic copy, then the only txn we're using was created by the connExecutor for us. We can only auto commit that txn when executing the INSERT pipeline for the final batch in implicit txn. If we don't auto-commit, then the connExecutor will do so for us (meaning we don't _have_ to, for correctness purposes). If we do what you're suggesting (and how the code used to be), then on second to last batch we performed some writes but couldn't auto commit, so the txn is still open. Then when we have `finalBatch = true`, but no rows to write, previously we would execute `preparePlannerForCopy` which would allow auto commit on the planner, but because `numRows` is zero below, we would short-circuit (so we wouldn't execute the INSERT pipeline when `planner.autoCommit` matters), so we would go straight to the deferred `cleanup` function in which we would do nothing because we have atomic copy (so the `if` condition is `false`).

Thus, from what I can tell the first commit in this PR has no difference in behavior for atomic copy, but it avoids creating an extra txn for non-atomic copy in case when `numRows` is zero. Does this match your reading of the code?


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Very nice!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-:-NnuVZrB0uSIyuWs26Xo:b-kerl87)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @rafiss)

___
*[`pkg/sql/copy_from.go` line 1090 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-NnohyZeDe9oGKd7iYVi:-NnuUiJFDJt0zA17stju:by5tr9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbd78f458ab20759553a6220ec589023c61e32eb/pkg/sql/copy_from.go#L1090)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, let me think.

If we have non-atomic copy, then we create a separate txn for each batch, and each txn is either auto-committed during the execution of the INSERT flow or in the deferred `cleanup` function returned by `preparePlannerForCopy` - in other words, it is effectively guaranteed (except for error cases) that if we called `preparePlannerForCopy` and it created a txn, then that txn will be committed.

If we have atomic copy, then the only txn we're using was created by the connExecutor for us. We can only auto commit that txn when executing the INSERT pipeline for the final batch in implicit txn. If we don't auto-commit, then the connExecutor will do so for us (meaning we don't _have_ to, for correctness purposes). If we do what you're suggesting (and how the code used to be), then on second to last batch we performed some writes but couldn't auto commit, so the txn is still open. Then when we have `finalBatch = true`, but no rows to write, previously we would execute `preparePlannerForCopy` which would allow auto commit on the planner, but because `numRows` is zero below, we would short-circuit (so we wouldn't execute the INSERT pipeline when `planner.autoCommit` matters), so we would go straight to the deferred `cleanup` function in which we would do nothing because we have atomic copy (so the `if` condition is `false`).

Thus, from what I can tell the first commit in this PR has no difference in behavior for atomic copy, but it avoids creating an extra txn for non-atomic copy in case when `numRows` is zero. Does this match your reading of the code?
</blockquote></details>

I see, I didn't realize connExecutor would take care of committing if auto-commit didn't happen. Thanks!


<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

@rafiss do you mind giving this PR another look?

rafiss reviewed (approved):
very nice!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116512#-:-NoNoXgJ2UPUdpO7GCtq:biuimdb)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Thanks for the reviews! And thanks to Nathan for finding these issues!

bors r+
